### DISCUSSION TITLE
Dynamická inicializace skriptu pro React aplikaci

### DISCUSSION THREAD

**Discussion by lucienevludova-direct** [canarytrace/documentation#218] [RUM]
Ahoj, potřebovali bychom mít možnost programově inicializovat skript, aby se mohl spustit kdykoliv během životního cyklu aplikace.
 
V kódu vidíme, že se vše potřebné pro inicializaci skriptu spouští při "load" události
```
window.addEventListener("load", (()=>{
    payload = structuredClone(base),
    collector()
}
```
Potřebovali bychom, aby to šlo spustit také programově (případně další metody, které poslouchají na "load"). Jinými slovy, abychom mohli kdykoliv zavolat něco jako CRUM.init(). Jinak to na Reactu a nejspíš ani v jiné SPA nebude fungovat, protože dnes máme jen dvě možnosti:
- spustit skript co nejdřív, takže před tím než se stránka v Reactu kompletně vyrenderuje - pak zafunguje "load" event, ale nemůžeme sbírat crum-action, protože ty prvky v DOM ještě neexistují a tak se to na ně nepověsí,
- spustit skript až po kompletním načtení, ale pak se nikdy nespustí "load" event, protože už nastal před spuštěním skriptu.

Děkujeme a přejeme pěkný den :)

**Comment by rdpanek** [comment]
Ahoj @lucienevludova-direct RUM se běžně používá na app napsané v reactu, next, vue dokonce i angluar včetně spa :) Co je přesně myšleno tím: spustit skript co nejdřív, takže před tím než se stránka v Reactu kompletně vyrenderuje - pak zafunguje "load" event, ale nemůžeme sbírat crum-action, protože ty prvky v DOM ještě neexistují a tak se to na ně nepověsí.

O tom to přesně ale je. Co nejdříve spustit RUM a pak se donačítá zbytek stránky, včetně initu stránky/frameworku a proto v RUM běží iterace po nastavených vteřinách, aby se sebrala nová data / objekty, která budou k dispozici až je stránka vyrenderuje / spustí.

Asi spíše neorzumím tomu bodu, který jsi uvedla, tak prosím o upřesnění / vysvětlení. 

Díky

**Comment by JanMachacek-Direct** [comment]
Dobrý den, @rdpanek, volali jsme spolu na konci července. Problém máme vlastně jen s tím crum-action atributem. Vidíme, že nám RUM běží co 3 vteřiny, ale ačkoliv elementy s tímto atributem v DOM máme, tak se nám to pak ve výsledcích nijak neprojevuje: pole v objektu view.actions v konzoli při ?debug=dry je vždy prázdné a nepropisuje se to ani nikam dál, třeba do Kibany (ačkoliv přesně nevím, jak nám tato integrace má fungovat). Platí tedy, že pokud se v DOM vyskytne nový prvek s crum-action atributem (řekněme že se tam objeví po stisku nějakého tlačítka), RUMko si ho načte v rámci 3s intervalu? Když si debugguji zdrojáky RUMka, tak se mi elementy atributem crum-action procházejí jen při load eventě, kdy se inicializuje skript a pak už ne. Z toho důvodu se bez problémů spustí RUMko, každé 3s se odesílají data - to je OK. Elementy s crum-action tou dobou ale ještě nejsou vygenerované na stránce, to se stane až o chvíli později a to si myslím, že způsobuje tyto potíže. Děkuji za radu, jak to správně nastavit.

**Reply by rdpanek** [reply]
Dobrý den, děkuji a teď už tomu rozumím a i Lucce. Ano, `crum-action` se inicializuje pouze při první smyčce, což pro tento případ nestačí a chtělo by to, aby se crum-action handloval během celého životního cyklu dokumentu.

Alternativně lze použít `CRUM.addEvent()` http://v2doc.canarytrace.com/docs/rum/rumClient#crum-functions což ale není to samé, jako `crum-action` a přijde mi použití `CRUM.addEvent()` v tomto případě jako "rovnák na vohejbák" což z mého pohledu není optimální.

Uděláme upgrade, ve kterém tohle pořešíme (pravděpodobně kontinuální sbírání crum-* akcí) jen se to bude muset proměřit na našich (mnoho) elementových webech, jestli to nehne s performance (což by nemělo, protože to není addEventListener na *) ale tak pro jistotu.

Uděláme to co nejdříve a napíšu. Díky

**Reply by rdpanek** [reply]
Dobrý den @JanMachacek-Direct aktualizace je v tomto docker image `docker pull quay.io/canarytrace/rum:2.10.2` zkuste prosím. cc @lucienevludova-direct a napište prosím, jestli je to nyní ok. Díky

**Reply by lucienevludova-direct** [reply]
Ahoj, budeme to ještě více zkoušet, ale na první pohled to vypadá, že úpravy zafungovaly a action se nám sbírá :+1:
